airflow/jobs/base_job.py

```diff
@@ -147,7 +147,7 @@ def heartbeat_callback(self, session=None):
         Callback that is called during heartbeat. This method should be overwritten.
         """
 
-    def heartbeat(self):
+    def heartbeat(self, only_if_necessary=False):
```

Comment:
nitpick, would be nice to add type hint here.

Reply:
Not sure why we need the `only_if_necessary` flag. Is it just for testing purpose?

Reply:
This allows heartbeat to called more frequently than the interval, but without it sleeping un-ncessarily.

This was (at the time) a performance improvement for the scheduler. Perhaps I should change it to `sleep=False`

Reply:
Oh no, that isn't what it does. Let me double check what sort of % performance difference this makes to scheduling. I think the name is right for what it does.

Reply:
> nitpick, would be nice to add type hint here.

Is that needed given it has a default?

Reply:
yes, having type hint prevents caller from passing in with a variable of wrong type by accident. it will also help with type checking within this method anywhere `only_if_necessary` is used.